javaslang/src/main/java/javaslang/collection/AbstractsQueue.java

```diff
@@ -0,0 +1,100 @@
+package javaslang.collection;
+
+import javaslang.*;
+import javaslang.control.Option;
+
+import java.util.*;
+
+abstract class AbstractsQueue<T, Q extends AbstractsQueue<T, Q>> implements Kind1<Q, T>, Traversable<T> {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Removes an element from this Queue.
+     *
+     * @return a tuple containing the first element and the remaining elements of this Queue
+     * @throws NoSuchElementException if this Queue is empty
+     */
+    abstract Tuple2<T, Q> dequeue();
```

Comment:
I think we can implement it here. However, `Queue.dequeue` does perform an extra empty-check, `PriorityQueue.dequeue` does not. In general we do these extra checks in order

1. to be independent of the methods that are called
2. to provide more details error messages what did go wrong

Therefore let's take the Queue method:

```java
    @Override
    public Tuple2<T, Q> dequeue() {
        if (isEmpty()) {
            throw new NoSuchElementException("dequeue of empty Queue");
        } else {
            return Tuple.of(head(), tail());
        }
    }
```

Maybe there are more Queue-specific methods that can be moved to this abstract class? Many methods of AbstractMap are also implementations of Traversable methods. We should move all duplicate methods from Queue and PriorityQueue to AbstractQueue.